File: Sources/FLACintosh/Library/JellyfinClient.swift
import Foundation

/// Jellyfin.
///
/// Unlike Subsonic there is a login step: username and password are
/// exchanged once for an access token, and every later request carries it.
actor JellyfinClient: MusicServerClient {
    let server: MusicServer
    private let password: String
    private let session: URLSession

    private var token: String?
    private var userID: String?

    /// Identifies this client to the server; Jellyfin shows it in its
    /// devices list, so it should say what it actually is.
    private static let deviceID = "macos-music-player"

    init(server: MusicServer, password: String, session: URLSession = .shared) {
        self.server = server
        self.password = password
        self.session = session
    }

    func albums(progress: @escaping @Sendable (_ done: Int, _ total: Int) -> Void) async throws -> [LibraryAlbum] {
        let user = try await authenticate()
        let albums: ItemsResponse = try await get(
            "Items",
            [
                "userId": user,
                "IncludeItemTypes": "MusicAlbum",
                "Recursive": "true",
                "SortBy": "SortName",
                "Fields": "DateCreated,ProductionYear,AlbumArtist",
                "Limit": "5000",
            ]
        )

        let listed = albums.Items ?? []
        progress(0, listed.count)
        var found: [LibraryAlbum] = []
        for (index, album) in listed.enumerated() {
            found.append(try await detail(album, user: user))
            progress(index + 1, listed.count)
        }
        return found
    }

    private func detail(_ album: Item, user: String) async throws -> LibraryAlbum {
        let songs: ItemsResponse = try await get(
            "Items",
            [
                "userId": user,
                "ParentId": album.Id,
                "IncludeItemTypes": "Audio",
                "SortBy": "ParentIndexNumber,IndexNumber,SortName",
                "Fields": "RunTimeTicks",
                "Limit": "500",
            ]
        )

        let artist = album.AlbumArtist ?? album.Name ?? "Unknown Artist"
        let tracks = (songs.Items ?? []).enumerated().map { index, song in
            LibraryTrack(
                id: streamURL(for: song.Id) ?? server.address,
                title: song.Name ?? "Untitled",
                artist: song.AlbumArtist ?? artist,
                albumArtist: artist,
                album: album.Name ?? "Unknown Album",
                trackNumber: song.IndexNumber ?? index + 1,
                discNumber: song.ParentIndexNumber,
                // Jellyfin counts in ticks: ten million to the second.
                duration: song.RunTimeTicks.map { Double($0) / 10_000_000 },
                hasLyrics: false,
                artworkURL: coverURL(album.Id, maxHeight: 600),
                source: .server(server.id)
            )
        }

        return LibraryAlbum(
            id: "\(server.id.uuidString)|\(album.Id)",
            title: album.Name ?? "Unknown Album",
            artist: artist,
            tracks: tracks,
            cover: try? await cover(album.Id),
            addedAt: album.DateCreated.flatMap { ISO8601DateFormatter().date(from: $0) } ?? .distantPast,
            year: album.ProductionYear.map(String.init),
            source: .server(server.id)
        )
    }

    func cover(_ albumID: String) async throws -> Data? {
        guard let url = coverURL(albumID, maxHeight: 320) else { return nil }
        let (data, response) = try await session.data(from: url)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }
        return data
    }

    /// Images need no token: Jellyfin serves them to anyone who can reach it.
    nonisolated private func coverURL(_ albumID: String, maxHeight: Int) -> URL? {
        var components = URLComponents(
            url: server.address.appendingPathComponent("Items/\(albumID)/Images/Primary"),
            resolvingAgainstBaseURL: false
        )
        components?.queryItems = [URLQueryItem(name: "maxHeight", value: String(maxHeight))]
        return components?.url
    }

    nonisolated func streamURL(for trackID: String) -> URL? {
        // The token goes in the query rather than a header: this URL is
        // handed to the downloader, which has no idea it is talking to
        // Jellyfin.
        guard let token = cachedToken else { return nil }
        var components = URLComponents(
            url: server.address.appendingPathComponent("Audio/\(trackID)/stream"),
            resolvingAgainstBaseURL: false
        )
        components?.queryItems = [
            URLQueryItem(name: "static", value: "true"),
            URLQueryItem(name: "api_key", value: token),
        ]
        return components?.url
    }

    /// Read from a non-isolated cache so `streamURL` can stay synchronous —
    /// it is called from view code building a track list.
    nonisolated private var cachedToken: String? { TokenCache.shared.token(for: server.id) }

    func ping() async throws {
        _ = try await authenticate()
    }

    /// Asks Jellyfin to look for new files now rather than at its next
    /// scheduled scan — after a download, so the record shows up in minutes
    /// instead of hours. Only an administrator may; anyone else is refused,
    /// and the library still reloads later on its own.
    func refreshLibrary() async throws {
        _ = try await authenticate()
        var request = URLRequest(url: server.address.appendingPathComponent("Library/Refresh"))
        request.httpMethod = "POST"
        request.setValue(authorizationHeader, forHTTPHeaderField: "Authorization")

        let response: URLResponse
        do {
            (_, response) = try await session.data(for: request)
        } catch {
            throw MusicServerError.notReachable
        }
        guard let http = response as? HTTPURLResponse, (200 ..< 300).contains(http.statusCode) else {
            throw MusicServerError.badResponse("Jellyfin did not start a library scan")
        }
    }

    // MARK: - Auth

    @discardableResult
    private func authenticate() async throws -> String {
        if let userID, token != nil { return userID }

        var request = URLRequest(url: server.address.appendingPathComponent("Users/AuthenticateByName"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(authorizationHeader, forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(
            AuthRequest(Username: server.username, Pw: password)
        )

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw MusicServerError.notReachable
        }
        guard let http = response as? HTTPURLResponse else { throw MusicServerError.notReachable }
        guard http.statusCode != 401 else { throw MusicServerError.auth }
        guard http.statusCode == 200 else {
            throw MusicServerError.badResponse("Login returned \(http.statusCode)")
        }

        let result = try JSONDecoder().decode(AuthResponse.self, from: data)
        token = result.AccessToken
        userID = result.User.Id
        TokenCache.shared.set(result.AccessToken, for: server.id)
        return result.User.Id
    }

    /// Sent as `Authorization`. Jellyfin 10.11 and later answer the legacy
    /// `X-Emby-Authorization` / `X-MediaBrowser-Token` headers with 400 and
    /// 401 unless the server re-enables legacy authorization.
    private var authorizationHeader: String {
        var header = """
        MediaBrowser Client="macos-music-player", Device="Mac", \
        DeviceId="\(Self.deviceID)", Version="0.1"
        """
        if let token { header += ", Token=\"\(token)\"" }
        return header
    }

    // MARK: - Requests

    private func get<T: Decodable>(_ path: String, _ parameters: [String: String]) async throws -> T {
        let token = try await authenticate() // ensures a token exists
        _ = token

        var components = URLComponents(
            url: server.address.appendingPathComponent(path),
            resolvingAgainstBaseURL: false
        )
        components?.queryItems = parameters.map { URLQueryItem(name: $0.key, value: $0.value) }
        guard let url = components?.url else { throw MusicServerError.notReachable }

        var request = URLRequest(url: url)
        request.setValue(authorizationHeader, forHTTPHeaderField: "Authorization")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw MusicServerError.notReachable
        }
        guard let http = response as? HTTPURLResponse else { throw MusicServerError.notReachable }
        guard http.statusCode != 401 else { throw MusicServerError.auth }
        guard http.statusCode == 200 else {
            throw MusicServerError.badResponse("\(path) returned \(http.statusCode)")
        }
        return try JSONDecoder().decode(T.self, from: data)
    }

    // MARK: - Wire types

    private struct AuthRequest: Encodable {
        let Username: String
        let Pw: String
    }

    private struct AuthResponse: Decodable {
        let AccessToken: String
        let User: User

        struct User: Decodable { let Id: String }
    }

    private struct ItemsResponse: Decodable {
        var Items: [Item]?
    }

    private struct Item: Decodable {
        var Id: String
        var Name: String?
        var AlbumArtist: String?
        var ProductionYear: Int?
        var DateCreated: String?
        var IndexNumber: Int?
        var ParentIndexNumber: Int?
        var RunTimeTicks: Int?
    }
}

/// Access tokens, so a stream URL can be built without awaiting an actor.
final class TokenCache: @unchecked Sendable {
    static let shared = TokenCache()

    private let lock = NSLock()
    private var tokens: [UUID: String] = [:]

    func token(for id: UUID) -> String? {
        lock.lock()
        defer { lock.unlock() }
        return tokens[id]
    }

    func set(_ token: String, for id: UUID) {
        lock.lock()
        defer { lock.unlock() }
        tokens[id] = token
    }
}

File: Windows/src/MusicPlayerWin.Core/Servers/JellyfinClient.cs
using System.Net.Http.Json;
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using MusicPlayerWin.Core.Library;

namespace MusicPlayerWin.Core.Servers;

public sealed class JellyfinClient : IMusicServerClient
{
    private readonly MusicServer _server;
    private readonly string _password;
    private readonly HttpClient _http;
    private readonly SemaphoreSlim _authGate = new(1, 1);
    private string? _token;
    private string? _userId;

    public JellyfinClient(MusicServer server, string password, HttpClient? httpClient = null)
    {
        _server = server;
        _password = password;
        _http = httpClient ?? new HttpClient { Timeout = TimeSpan.FromSeconds(60) };
    }

    public async Task<IReadOnlyList<LibraryAlbum>> AlbumsAsync(Action<int, int>? progress = null, CancellationToken cancellationToken = default)
    {
        var user = await AuthenticateAsync(cancellationToken).ConfigureAwait(false);
        var response = await GetAsync<ItemsResponse>("Items", new Dictionary<string, string>
        {
            ["userId"] = user,
            ["IncludeItemTypes"] = "MusicAlbum",
            ["Recursive"] = "true",
            ["SortBy"] = "SortName",
            ["Fields"] = "DateCreated,ProductionYear,AlbumArtist",
            ["Limit"] = "5000"
        }, cancellationToken).ConfigureAwait(false);

        var items = response.Items ?? [];
        progress?.Invoke(0, items.Count);
        var result = new List<LibraryAlbum>(items.Count);
        for (var i = 0; i < items.Count; i++)
        {
            cancellationToken.ThrowIfCancellationRequested();
            result.Add(await ReadAlbumAsync(items[i], user, cancellationToken).ConfigureAwait(false));
            progress?.Invoke(i + 1, items.Count);
        }
        return result;
    }

    public async Task<byte[]?> CoverAsync(string albumId, CancellationToken cancellationToken = default)
    {
        var url = Build("Items/" + Uri.EscapeDataString(albumId) + "/Images/Primary", new Dictionary<string, string> { ["maxHeight"] = "320" });
        try
        {
            using var response = await _http.GetAsync(url, cancellationToken).ConfigureAwait(false);
            return response.StatusCode == HttpStatusCode.OK ? await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false) : null;
        }
        catch (HttpRequestException ex)
        {
            throw new MusicServerException("Could not reach Jellyfin.", ex);
        }
    }

    public Uri? StreamUri(string trackId)
    {
        if (string.IsNullOrWhiteSpace(_token)) return null;
        return Build($"Audio/{Uri.EscapeDataString(trackId)}/stream", new Dictionary<string, string>
        {
            ["static"] = "true",
            ["api_key"] = _token
        });
    }

    public async Task<IReadOnlyList<ServerPlaylist>> PlaylistsAsync(CancellationToken cancellationToken = default)
    {
        var user = await AuthenticateAsync(cancellationToken).ConfigureAwait(false);
        var listed = await GetAsync<ItemsResponse>("Items", new Dictionary<string, string>
        {
            ["userId"] = user,
            ["IncludeItemTypes"] = "Playlist",
            ["Recursive"] = "true",
            ["Fields"] = "MediaType"
        }, cancellationToken).ConfigureAwait(false);

        var result = new List<ServerPlaylist>();
        foreach (var playlist in listed.Items ?? [])
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (playlist.MediaType is not null && !string.Equals(playlist.MediaType, "Audio", StringComparison.OrdinalIgnoreCase))
                continue;

            var items = await GetAsync<ItemsResponse>($"Playlists/{Uri.EscapeDataString(playlist.Id)}/Items",
                new Dictionary<string, string> { ["userId"] = user }, cancellationToken).ConfigureAwait(false);

            var urls = (items.Items ?? []).Select(x => StreamUri(x.Id)).Where(x => x is not null).Cast<Uri>().ToList();
            if (urls.Count == 0) continue;
            result.Add(new ServerPlaylist($"{_server.Id:N}/{playlist.Id}", playlist.Name ?? "Playlist", urls, _server.Id) { ServerName = _server.Name });
        }
        return result;
    }

    public async Task PingAsync(CancellationToken cancellationToken = default)
    {
        _ = await AuthenticateAsync(cancellationToken).ConfigureAwait(false);
    }

    private async Task<LibraryAlbum> ReadAlbumAsync(Item album, string user, CancellationToken ct)
    {
        var songs = await GetAsync<ItemsResponse>("Items", new Dictionary<string, string>
        {
            ["userId"] = user,
            ["ParentId"] = album.Id,
            ["IncludeItemTypes"] = "Audio",
            ["SortBy"] = "ParentIndexNumber,IndexNumber,SortName",
            ["Fields"] = "RunTimeTicks",
            ["Limit"] = "500"
        }, ct).ConfigureAwait(false);

        var artist = album.AlbumArtist ?? album.Name ?? "Unknown Artist";
        var trackList = (songs.Items ?? []).Select((song, index) => new LibraryTrack
        {
            Id = StreamUri(song.Id) ?? _server.Address,
            Title = song.Name ?? "Untitled",
            Artist = song.AlbumArtist ?? artist,
            AlbumArtist = artist,
            Album = album.Name ?? "Unknown Album",
            TrackNumber = song.IndexNumber ?? index + 1,
            DiscNumber = song.ParentIndexNumber,
            Duration = song.RunTimeTicks.HasValue ? song.RunTimeTicks.Value / 10_000_000d : null,
            HasLyrics = false,
            ArtworkUrl = Build($"Items/{Uri.EscapeDataString(album.Id)}/Images/Primary", new Dictionary<string, string> { ["maxHeight"] = "600" }),
            Source = LibrarySource.Server(_server.Id)
        }).ToList();

        return new LibraryAlbum
        {
            Id = $"{_server.Id:N}|{album.Id}",
            Title = album.Name ?? "Unknown Album",
            Artist = artist,
            Tracks = trackList,
            Cover = await CoverAsync(album.Id, ct).ConfigureAwait(false),
            AddedAt = ServerDate.Parse(album.DateCreated) ?? DateTimeOffset.MinValue,
            Year = album.ProductionYear?.ToString(),
            Source = LibrarySource.Server(_server.Id)
        };
    }

    private async Task<string> AuthenticateAsync(CancellationToken cancellationToken)
    {
        if (_userId is not null && _token is not null) return _userId;

        await _authGate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            if (_userId is not null && _token is not null) return _userId;

            using var request = new HttpRequestMessage(HttpMethod.Post, Build("Users/AuthenticateByName", new Dictionary<string, string>()));
            request.Headers.TryAddWithoutValidation("Authorization", AuthorizationHeader());
            request.Content = new StringContent(JsonSerializer.Serialize(new { Username = _server.Username, Pw = _password }), Encoding.UTF8, "application/json");

            HttpResponseMessage response;
            try { response = await _http.SendAsync(request, cancellationToken).ConfigureAwait(false); }
            catch (HttpRequestException ex) { throw new MusicServerException("Could not reach Jellyfin.", ex); }

            if (response.StatusCode == HttpStatusCode.Unauthorized) throw new MusicServerException("Jellyfin rejected the credentials.");
            if (!response.IsSuccessStatusCode) throw new MusicServerException($"Jellyfin login returned {(int)response.StatusCode}.");

            var auth = await response.Content.ReadFromJsonAsync<AuthResponse>(cancellationToken: cancellationToken).ConfigureAwait(false)
                       ?? throw new MusicServerException("Jellyfin returned an empty login response.");
            _token = auth.AccessToken;
            _userId = auth.User.Id;
            return _userId;
        }
        finally
        {
            _authGate.Release();
        }
    }

    private async Task<T> GetAsync<T>(string path, IReadOnlyDictionary<string, string> query, CancellationToken ct)
    {
        var user = await AuthenticateAsync(ct).ConfigureAwait(false);
        _ = user;
        using var request = new HttpRequestMessage(HttpMethod.Get, Build(path, query));
        request.Headers.TryAddWithoutValidation("Authorization", AuthorizationHeader());
        HttpResponseMessage response;
        try { response = await _http.SendAsync(request, ct).ConfigureAwait(false); }
        catch (HttpRequestException ex) { throw new MusicServerException("Could not reach Jellyfin.", ex); }
        if (response.StatusCode == HttpStatusCode.Unauthorized) throw new MusicServerException("Jellyfin authentication expired.");
        if (!response.IsSuccessStatusCode) throw new MusicServerException($"Jellyfin {path} returned {(int)response.StatusCode}.");
        return await response.Content.ReadFromJsonAsync<T>(cancellationToken: ct).ConfigureAwait(false)
               ?? throw new MusicServerException($"Jellyfin {path} returned no usable JSON.");
    }

    private Uri Build(string path, IReadOnlyDictionary<string, string> query)
    {
        var baseAddress = new Uri(_server.Address.ToString().TrimEnd('/') + "/");
        var target = new Uri(baseAddress, path.TrimStart('/'));
        return ServerUrl.WithQuery(target, query);
    }

    private string AuthorizationHeader()
    {
        var header = $"MediaBrowser Client=\"MusicPlayerWin\", Device=\"Windows\", DeviceId=\"musicplayerwin\", Version=\"0.2\"";
        if (_token is not null) header += $", Token=\"{_token}\"";
        return header;
    }

    private sealed record AuthResponse(string AccessToken, AuthUser User);
    private sealed record AuthUser(string Id);
    private sealed record ItemsResponse(List<Item>? Items);
    private sealed record Item
    {
        public required string Id { get; init; }
        public string? Name { get; init; }
        public string? AlbumArtist { get; init; }
        public int? ProductionYear { get; init; }
        public string? DateCreated { get; init; }
        public int? IndexNumber { get; init; }
        public int? ParentIndexNumber { get; init; }
        public long? RunTimeTicks { get; init; }
        public string? MediaType { get; init; }
    }
}
